test/jdk/java/lang/invoke/TestVHInvokerCaching.java

```diff
@@ -0,0 +1,96 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8265079
+ * @run testng/othervm -Xverify:all -ea -esa TestVHInvokerCaching
+ */
+
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.lang.invoke.MethodHandles.lookup;
+import static org.testng.Assert.assertSame;
+
+public class TestVHInvokerCaching {
+
+    @Test(dataProvider = "testHandles")
+    public static void testVHIInvokerCaching(VarHandle testHandle) throws Throwable {
+        for (VarHandle.AccessMode mode : VarHandle.AccessMode.values()) {
+            MethodHandle handle1 = MethodHandles.varHandleInvoker(mode, testHandle.accessModeType(mode));
+            MethodHandle handle2 = MethodHandles.varHandleInvoker(mode, testHandle.accessModeType(mode));
+            assertSame(handle1, handle2);
+        }
+
+        for (VarHandle.AccessMode mode : VarHandle.AccessMode.values()) {
+            MethodHandle handle1 = MethodHandles.varHandleExactInvoker(mode, testHandle.accessModeType(mode));
+            MethodHandle handle2 = MethodHandles.varHandleExactInvoker(mode, testHandle.accessModeType(mode));
+            assertSame(handle1, handle2);
+        }
+    }
+
+    @DataProvider
+    public static Object[][] testHandles() throws NoSuchFieldException, IllegalAccessException {
+        List<VarHandle> testHandles = new ArrayList<>();
+
+        Class<?>[] TEST_TYPES = {
+            byte.class,
+            short.class,
+            char.class,
+            int.class,
+            long.class,
+            float.class,
+            double.class,
+            Object.class,
+            String.class
+        };
+
+        class Holder {
+            byte f_byte;
+            short f_short;
+            char f_char;
+            int f_int;
+            long f_long;
+            float f_float;
+            double f_double;
+            Object f_Object;
+            String f_String;
+        }
+
+        MethodHandles.Lookup lookup = lookup();
+
+        for (Class<?> type : TEST_TYPES) {
```

Comment:
A simpler approach would be to iterate over the fields of `Holder` and use `unreflectVarHandle`, then you can remove `TEST_TYPES`.

Reply:
Yeah, that's a good idea. Fixed